### PR TITLE
Adding support for attaching custom params to an airbrake notice from an error event

### PR DESCRIPTION
**Pull request by softwarebygabe**
## PR Summary

*Replace this text with a description of what this PR is accomplishing*

 - **Type:** Features
 - **Intended Change Level:** patch

#### Reviewers:

Reviewers keep the following in mind while reviewing this PR, and provide an assessment of them in your approval comment:

- Does the "Intended Change Level" above match the actual behavior of this PR?
- Is this PR following patterns set forth in the package (if modifying a package), or the go-sdk design patterns if implementing a new package from scratch?
- Does this require a backport to LTS versions? If so ping, let the contributors group know.

### CODE REVIEW

File: airbrake/notice.go [hunk 1]
```diff
@@ -33,7 +36,7 @@ func NewNotice(err interface{}, req *http.Request) *gobrake.Notice {
 			Context: make(map[string]interface{}),
 			Env:     make(map[string]interface{}),
 			Session: make(map[string]interface{}),
-			Params:  make(map[string]interface{}),
+			Params:  make(Params),
```
Comment (wcharczuk) [inline airbrake/notice.go]: if we do want to keep this as `map string=>interface{}` we'll want to add a field and an option mutator for error events in the logger.

File: airbrake/notice.go [hunk 2]
```diff
@@ -9,6 +9,9 @@ import (
 	"github.com/blend/go-sdk/webutil"
 )
 
+// Params is a custom type alias
+type Params map[string]interface{}
```
Comment (wcharczuk) [inline airbrake/notice.go]: why the type alias?

File: airbrake/notice.go [hunk 3]
```diff
@@ -64,6 +67,13 @@ func NewNotice(err interface{}, req *http.Request) *gobrake.Notice {
 	return notice
 }
 
+// NewNoticeWithParams adds a map of params to the new default gobrake notice.
+func NewNoticeWithParams(err interface{}, params map[string]interface{}) *gobrake.Notice {
```
Comment (dhermes) [inline airbrake/notice.go]: `map[string]interface{}` or your `Params` alias here?

Comment (softwarebygabe) [reply airbrake/notice.go]: good catch, got rid of the alias now

File: airbrake/listener.go [hunk 4]
```diff
@@ -21,6 +21,11 @@ func AddListeners(log logger.Listenable, cfg Config) {
 	listener := logger.NewErrorEventListener(func(_ context.Context, ee *logger.ErrorEvent) {
 		if req, ok := ee.State.(*http.Request); ok {
 			client.NotifyWithRequest(ee.Err, req)
+		} else if state, ok := ee.State.(interface{}); ok {
+			errStateMap := map[string]interface{}{
+				"ErrorState": state,
+			}
```
Comment (softwarebygabe) [inline airbrake/listener.go]: how do we feel about this approach?

File: airbrake/notice.go (hunk 2 above)
Comment (softwarebygabe) [reply airbrake/notice.go]: rm'd

File: airbrake/listener.go (hunk 4 above)
Comment (softwarebygabe) [reply airbrake/listener.go]: basically now just passing anything on the error state unless it is of the specific http request type

**Comment by softwarebygabe** [comment]
Now the API, if you wanted to airbrake custom params from the logger would look like:
```
// add airbrake listeners
if !cfg.Airbrake.IsZero() {
  airbrake.AddListeners(log, cfg.Airbrake)
}

// airbrake something
log.ErrorWithState(errors.New("destined for airbrake"), map[string]string{
  "some_relevant_key": "the_value",
})
```

File: airbrake/listener.go (hunk 4 above)
Comment (wcharczuk) [reply airbrake/listener.go]: to put it bluntly, the type assertion `.(interface{})` is pointless, everything ascribes to the empty interface, just test if it's != nil

File: logger/context.go [hunk 5]
```diff
@@ -142,14 +142,18 @@ func (sc Context) Error(err error) error {
 	return err
 }
 
-// ErrorWithReq logs an error to std err with a request.
-func (sc Context) ErrorWithReq(err error, req *http.Request) error {
-	ee := NewErrorEvent(Error, err)
-	ee.State = req
+// ErrorWithState logs an error to std with a generic state.
+func (sc Context) ErrorWithState(err error, state interface{}) error {
```
Comment (wcharczuk) [inline logger/context.go]: i wouldn't have both this and ErrorWithReq, i would have an `...ErrorOption` variadic param on `Error()`, `Warning()` and `Fatal()` that mutates the error event object, setting the state, or setting a request property.

**Comment by wcharczuk** [comment]
@softwarebygabe hey we're going to close this out in favor of the sentry tooling we're going to add soon. will circle back with you on this after.

**Pull request closed without merging**